testutils/src/main/java/com/microsoft/identity/internal/testutils/CacheUtils.java

```diff
@@ -0,0 +1,147 @@
+        for (Map.Entry<String, ?> cacheEntry : cacheEntries.entrySet()) {
+            String keyToEdit = cacheEntry.getKey();
+            if (predicate.test(keyToEdit)) {
+                Object cacheValue = cacheEntries.get(keyToEdit);
+                if (cacheValue instanceof String) {
+                    Credential credential = CACHE_KEY_VALUE_DELEGATE.fromCacheValue((String) cacheValue, Credential.class);
+                    credential.setSecret(editor.apply(credential.getSecret()));
+                    prefEditor.putString(keyToEdit, CACHE_KEY_VALUE_DELEGATE.generateCacheValue(credential));
+                    prefEditor.apply();
+                }
+            }
+        }
+    }
+
+    /**
+     * utility function to edit Access Token.
+     *
+     * @param sharedPrefName Name of the shared preference.
+     * @param editor         Functional interface for token editing method.
+     */
+    public void editAllAccessTokenInCache(@NonNull final String sharedPrefName, Function<String, String> editor) {
+        Predicate<String> predicate = new Predicate<String>() {
+            @Override
+            public boolean test(String cacheKey) {
+                return TestUtils.isAccessToken(cacheKey);
+            }
+        };
+
+        editAllTokenInCache(sharedPrefName, predicate, editor);
+    }
+
+    /**
+     * utility function to edit Refresh Token.
+     *
+     * @param sharedPrefName Name of the shared preference.
+     * @param editor         Functional interface for token editing method.
+     */
+    public void editAllRefreshTokenInCache(@NonNull final String sharedPrefName, Function<String, String> editor) {
+        Predicate<String> predicate = new Predicate<String>() {
+            @Override
+            public boolean test(String cacheKey) {
+                return TestUtils.isRefreshToken(cacheKey);
+            }
+        };
+
+        editAllTokenInCache(sharedPrefName, predicate, editor);
+    }
+
+    /**
+     * Split the string into 3 segment separated by period "." and set the
+     * random character at the random position in the 3rd segment.
+     *
+     * @param string string to be edited.
+     * @return edited string.
+     */
+    private static String editTokenSignature(String string) {
+        String[] segments = string.split(".");
+        if (segments.length != 2) {
+            throw new AssertionError("not JWT");
+        }
+
+        String signature = segments[segments.length - 1];
+        StringBuilder sb = new StringBuilder(signature);
+        Random rnd = new Random();
+        int rndIndex = rnd.nextInt(sb.length());
+        char charAtRndPosition = sb.charAt(rndIndex);
+        char rndChar = (char) ('a' + rnd.nextInt(26));
+
+        if (charAtRndPosition == rndChar) {
+            rndChar = (char) (charAtRndPosition + rnd.nextInt(26));
+        }
+
+        sb.setCharAt(rndIndex, rndChar);
+        segments[segments.length - 1] = sb.toString();
```

Comment:
I have made a mistake here, there would be 3 segments and index 0,1,2 were segments of header, payload and signature respectively. I have updated and used index to avoid confusion.